crypto/src/lib.rs

```diff
@@ -274,89 +299,189 @@ impl From<KeyPair> for (PublicKey, PrivateKey) {
 
 ffi::ffi_item! {
     /// Public Key used in signatures.
-    #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, CopyGetters)]
-    #[cfg_attr(not(feature="ffi_import"), derive(DebugCustom, Display, Hash, DeserializeFromStr, SerializeDisplay, Decode, Encode, IntoSchema))]
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "self.normalize()"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "self.normalize()"))]
-    pub struct PublicKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        payload: ConstVec<u8>,
+    #[derive(Clone, PartialEq, Eq)]
+    #[cfg_attr(not(feature="ffi_import"), derive(DeserializeFromStr, SerializeDisplay))]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PublicKey {
+        Ed25519(ed25519::PublicKey),
+        Secp256k1(secp256k1::PublicKey),
+        BlsNormal(bls::BlsNormalPublicKey),
+        BlsSmall(Box<bls::BlsSmallPublicKey>),
     }
 }
 
 #[ffi_impl_opaque]
 impl PublicKey {
     /// Creates a new public key from raw bytes received from elsewhere
-    pub fn from_raw(algorithm: Algorithm, payload: ConstVec<u8>) -> Self {
-        Self {
-            digest_function: algorithm,
-            payload,
+    ///
+    /// # Errors
+    ///
+    /// Fails if public key parsing fails
+    pub fn from_raw(algorithm: Algorithm, payload: &[u8]) -> Result<Self, ParseError> {
+        match algorithm {
+            Algorithm::Ed25519 => {
+                ed25519::Ed25519Sha512::parse_public_key(payload).map(Self::Ed25519)
+            }
+            Algorithm::Secp256k1 => {
+                secp256k1::EcdsaSecp256k1Sha256::parse_public_key(payload).map(Self::Secp256k1)
+            }
+            Algorithm::BlsNormal => bls::BlsNormal::parse_public_key(payload).map(Self::BlsNormal),
+            Algorithm::BlsSmall => bls::BlsSmall::parse_public_key(payload)
+                .map(Box::new)
+                .map(Self::BlsSmall),
         }
     }
 
-    /// Extracts the raw bytes from public key
-    pub fn into_raw(self) -> (Algorithm, ConstVec<u8>) {
-        (self.digest_function, self.payload)
+    /// Extracts the raw bytes from public key, copying the payload.
+    ///
+    /// `into_raw()` without copying is not provided because underlying crypto
+    /// libraries do not provide move functionality.
+    pub fn to_raw(&self) -> (Algorithm, Vec<u8>) {
```

Comment:
why is it not called `into_raw` anymore? because it involves cloning?

Reply:
Yes

Reply:
look [here](https://doc.rust-lang.org/std/primitive.str.html#impl-From%3C%26str%3E-for-Box%3Cstr%3E). In `std` there is a `From<&str>` impl that involves cloning. It's obvious that cloning is happening when the receiver type is a reference.

Reply:
interesting